### DISCUSSION TITLE
Enabling Shake using BindingContext is a problem

### DISCUSSION THREAD

**Discussion by smurfster** [roubachof/MetroLog#28] [Q&A]
Hi,

Is there a method of globally enabling Shake as it seems strange to define at page level.
Also a bigger concern is that I have to make the BindingContext bind to LoggingController instead of a viewModel that provides the function to the page.

Am I missing something as Im pretty new to MAUI

**Comment by smurfster** [comment]
So the way I've implemented this is to have a base viewmodel that instantiates the LoggingController class

**Reply by ssuppan** [reply]
The shake & share aspect of this logger is what really attracted me to this product.  However, I have also implemented MVVM in .net maui and am having difficulty trying to get the "shake" to work.  If I change the binding context from my own view model to the LogController then I screw up my page and all of its bindings to my view model.

Perhaps Jean-Marie can create an MVVM example of how to use MetroLogger?  I apologize if I am missing something simple with respect to using it in MVVM but I cannot currently see it.

**Reply by roubachof** [reply]
Just instantiate the LogController in your view model and call the static methods `LogController.SuspendShake()` and `LogController.ResumeShakeIfNeeded()` when you need.
Or you can use the `ToggleShakeCommand` to enable/disable the shake to log.

**Reply by ssuppan** [reply]
Thank you Jean-Marie for answering so quickly.  
I was able to get it working as you described.  
However, I do not see the share logs button.  
Does that only work with the streaming file logger?

**Reply by ssuppan** [reply]
The only way I got this to work with MVVM was to inject the ILoggerFactory into my derived and base class.  From there, I create a logger using the derived classname as follows:

protected readonly ILogger _logger;
 
public BaseContentPageVM(ILoggerFactory loggerFactory)
{
  _logger = loggerFactory.CreateLogger(this.GetType().Name);
 }
 After this my _logger is available to both the base class and the derived class.  Still, not as clean or simple as passing in ILogger.

**Reply by roubachof** [reply]
You must have misconfigured your logger, look at how it's configured in the Maui sample

**Reply by ssuppan** [reply]
Jean-Marie, I just noticed that I can shake the device to popup the "share logs" page.  After I close it, a secondary shake DOES NOT popup the "share logs" page again.  Any suggestions?

My BaseContentPageVM's constructor creates a new instance of LogController:
            _logController = new LogController();
            _logController.IsShakeEnabled = true;
            
However, ResumeShakeIfNeeded() is not available from my new instance.

**Reply by ssuppan** [reply]
Jean-Marie, what might really help MetroLog usage is if you supplied an MVVM example as well as the code-behind.
I find that most .net maui demo's use only code-behind and leave it up to us to figure out how to work that into an MVVM solution.

**Reply by ssuppan** [reply]
Jean-Marie, I found a quick and dirty solution to my problem of popping up the "share logs" page from subsequent shakes.
My BaseContentPageVM has a ViewModelOnAppearing() method which gets called when the page is appearing.
Since this is called after popping down the "share logs" screen, I simply recreated the _logController and set IsShakeEnabled to true again:

        public virtual async Task ViewModelOnAppearing()
        {
            _logController = new LogController();
            _logController.IsShakeEnabled = true;
        }

**Reply by roubachof** [reply]
I don't think you need to instantiate a new LogController, you could just call `LogController.ResumeShakeIfNeeded()`

**Reply by ssuppan** [reply]
Jean-Marie, I tried that.  It didn't work.

**Reply by roubachof** [reply]
Oh i didn't see you did this at the view model level.
You should call `LogController.SuspendShake` when the App is stopped.
Then `LogController.ResumeShakeIfNeeded()` when it's resumed.
see: https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/app-lifecycle?view=net-maui-8.0#platform-lifecycle-events
It will be maybe neater :)
